sdk/core/src/services/GenericApiService.ts

```diff
@@ -0,0 +1,117 @@
+import keyBy from 'lodash.keyby'
+import FetchType from 'node-fetch'
+import { profile } from '@affinidi/common'
+
+import SdkError from '../shared/SdkError'
+import { GenericApiSpec, ExtractAllOperationIds, ExtractRequestType, ExtractResponseType } from './SwaggerTypes'
+
+let fetch: typeof FetchType
+
+/* istanbul ignore next */
+if (!fetch) {
+  fetch = require('node-fetch')
+}
+
+type BasicRequestOptions = {
+  authorization?: string
+  storageRegion?: string
+  urlPostfix?: string
+}
+
+type RequestOptions<TParams extends Record<string, any>> = TParams extends undefined
+  ? BasicRequestOptions
+  : BasicRequestOptions & { params: TParams }
+
+type ConstructorOptions = { accessApiKey: string }
+
+@profile()
+export default class GenericApiService<TApiSpec extends GenericApiSpec> {
+  private readonly _serviceUrl: string
+  private readonly _accessApiKey: string
+  private readonly _specGroupByOperationId
+
+  constructor(serviceUrl: string, options: ConstructorOptions, rawSpec: TApiSpec) {
+    this._serviceUrl = serviceUrl
+    this._accessApiKey = options.accessApiKey
+    const specGroupByOperationId = GenericApiService.parseSpec(rawSpec)
+    this._specGroupByOperationId = specGroupByOperationId
+  }
+
+  private static parseSpec<TApiSpec extends GenericApiSpec>(rawSpec: TApiSpec) {
+    const basePath = rawSpec.servers[0].url
+
+    const spec = Object.entries(rawSpec.paths).flatMap(([operationPath, operation]) => {
+      const path = `${basePath}${operationPath}`
+
+      return (['get', 'post', 'put', 'delete'] as const).flatMap((method) => {
+        const operationForMethod = operation[method]
+        if (!operationForMethod) {
+          return []
+        }
+
+        const { operationId }: { operationId: ExtractAllOperationIds<TApiSpec> } = operationForMethod
+
+        return [
+          {
+            path,
+            method,
+            operationId,
+          },
+        ]
+      })
+    })
+
+    return keyBy(spec, 'operationId') as Record<ExtractAllOperationIds<TApiSpec>, typeof spec[number]>
+  }
+
+  private static async executeByOptions<TResponse>(
+    accessApiKey: string,
+    method: string,
+    url: string,
+    options: BasicRequestOptions & { params?: unknown },
+  ) {
+    const { authorization, storageRegion, params, urlPostfix } = options
+    const fetchOptions = {
+      headers: {
+        Accept: 'application/json',
+        'Content-Type': 'application/json',
+        'Api-Key': accessApiKey,
+        ...(authorization && { Authorization: authorization }),
+        ...(storageRegion && { 'X-DST-REGION': storageRegion }),
```

Comment:
since we created a common api service, would it be a good idea to roll out `x-sdk-version` header with this version?

Reply:
I think it would.

Not in this version though; I have certain time constraints with other tasks now.